### DISCUSSION TITLE
Modifying the IDL generator to tweak the generated type support code

### DISCUSSION THREAD

**Discussion by mellegge** [jmmorato/openddsharp#175] [Q&A] [answered]
Is there a way to get the openddsharp.standard.idlgenerator to tweak its generated code? Specifically we would like to have the generated C# TypeSupport classes inherit from a common interface that contains the declarations for GetTypeName, RegisterType and UnregisterType.  This would make it possible to use generics when handling lots of TypeSupport classes from more complex IDL files.

More broadly, it would be nice if a config file could be supplied to the generator which would allow further tweaking but I fear that would be more complex.

Note, I am using the command line tools as per the getting started page.

Thanks
Mel

**Comment by jmmorato** [comment, marked as answer]
_Is there a way to get the openddsharp.standard.idlgenerator to tweak its generated code?_

Yes, the [CSharpImplTemplate.txt](https://github.com/jmmorato/openddsharp/blob/develop/Native/CSharpImplTemplate.txt) is copied in the `<IDL_NUGET_PATH>/OpenDDS_<arch>/dds/idl` folder. We could define interface in the project (i.e. ITypeSupport) and ensure `TypeSupport `classes inherit it in the code generator.

_This would make it possible to use generics when handling lots of TypeSupport classes from more complex IDL files._

Agree, and could be used for other implementations as well, so feel free to pull request your changes to be integrated in the next releases.

_More broadly, it would be nice if a config file could be supplied to the generator which would allow further tweaking but I fear that would be more complex._

Again agree, templating system is quite limited using the current implementation. I created a PoC some time ago to replace `tao_idl` by using an own code generator based in [ANTLR](https://www.antlr.org). I achieved a simple generator IDL to C# and IDL **from** C# (define C# classes in your project and provide the required IDL file to OpenDDS for compiling) integrated in a T4 templating system without a big effort but, as you foreseen, a full implementation will be more complex and requires a considerable amount of time. As a single developer with limited time for the project, I need to consider carefully the next features to implement but I'm pretty sure that in the future (don't ask me when) I will implement it.

EDIT: A better implementation would be to implement two interfaces, a generic `ITypeSupport` and a specific `I<TYPE_NAME>TypeSupport` to handle methods with undefined objects (`object` arguments) and specific IDL types objects. That's more or less the OpenDDS implementation where the specific `TypeSupport` classes inherits from the generic `::OpenDDS::DCPS::TypeSupport` class
